src/database.ts

```diff
@@ -1,389 +1,383 @@
-		const {_store: store, schema} = this;
-		return runAction<T | undefined>({
-			store,
-			schema,
-			mode: 'readwrite',
-			routine: (_store) =>
-				usingDatabase(store, () => handleRequest(_store.put(localItem), localItem))
-		});
-	}
-
-	private async updateRemote(item: T): Promise<T | undefined> {
-		if (!this.api) {
-			return item;
-		}
-		const isOnline = isOnlineSupport(this.api) ? await this.api.isOnline() : undefined;
-		if (isOnline === undefined || isOnline) {
-			return this.api.update(item).then(async (result: T | undefined) => {
-				const deleted = await this.getId(item, (id: IndexedKey) => this.deleteLocal(id, item)).catch(() => false);
-				return deleted && result ?  this.createLocal(result, false) : undefined;
-			}).catch((err) => {
-				console.error(`can't call update of api ${buildEntityIdent(this.schema)}`, err);
-				return item;
-			});
-		}
-		return item;
-	}
-
-	private deleteLocal(id: IndexedKey, item: T, addFlag: boolean = true): Promise<boolean> {
-		const {_store: store, schema} = this;
-		if (!addFlag) {
-			return runAction({
-				schema,
-				store,
-				mode: 'readwrite',
-				routine: (_store) =>
-					usingDatabase(store, () => handleRequest(_store.delete(id), item)
-							.then(() => true)
-							.catch(() => false))
-			});
-		}
-		const localItem: T & { flag?: string } = {...item, flag: 'D'};
-		return runAction<boolean>({
-			store,
-			schema,
-			mode: 'readwrite',
-			routine: (_store) => handleRequest(_store.put(localItem), true)
-				.then(((res) => res === undefined ? true : res))
-		});
-	}
-
-	private async deleteRemote(id: IndexedKey, item: T): Promise<boolean> {
-		if (!this.api) {
-			return false;
-		}
-		const isOnline = isOnlineSupport(this.api) ? await this.api.isOnline() : undefined;
-		if (isOnline === undefined || isOnline) {
-			return this.api.delete(item).then(async (deleted: boolean) => {
-				return deleted && await this.deleteLocal(id, item, false);
-			}).catch((err) => {
-				console.error(`can't call delete of api ${buildEntityIdent(this.schema)}`, err);
-				return false;
-			});
-		}
-		return true;
-	}
-
-	private getId<S>(item: T, callback: (id: IndexedKey) => Promise<S>): Promise<S> {
-		const {keyPath = 'id'} = this.schema;
-		if (!(item as any).hasOwnProperty(keyPath)) {
-			return Promise.reject();
-		}
-		const id: IndexedKey = getValue(item, keyPath as keyof T) as any;
-		return callback(id);
-	}
+  constructor(
+    private objectStore: IDBObjectStore,
+    private key: string,
+    private schema: StoreSchema,
+    private api?: CrudApi<T>
+  ) {}
+
+  /**
+   * push local transactions to remote and try to fetch updates from remote
+   */
+  public sync(): Promise<void> {
+    type F = WithFlag<T>;
+    return new Promise(async (done, _) => {
+      const isOnline = this.api && isOnlineSupport(this.api)
+        ? await this.api.isOnline()
+        : undefined;
+      if (isOnline === false) {
+        done();
+        return;
+      }
```

Comment:
```suggestion
  public async sync(): Promise<void> {
    type F = WithFlag<T>;
    const isOnline = this.api && isOnlineSupport(this.api)
      ? await this.api.isOnline()
      : true;
    if (isOnline === false) {
      return;
    }
    return new Promise(async (done, _) => {
```

Custom promisification is difficult. Ideally, most code is outside of the promise constructor to prevent mistakes silently being ignored.